docs/tutorials/auction.md

```diff
@@ -0,0 +1,244 @@
+
+### Close
+
+Method `public close(sig: Sig)` is simple, we require:
+
+- It can only be called by the auctioneer. That is why we need to pass in the caller's signature.
+
+```ts
+// Check signature of the auctioneer.
+assert(this.checkSig(sig, this.auctioneer), 'signature check failed')
+```
+
+- Now the auction deadline has passed
+
+```ts
+assert(this.ctx.locktime >= this.auctionDeadline, 'auction is not over yet')
+```
+
+:::note
+We don't place any constraint on transaction outputs here, because the auctioneer can send the highest bid to any address he controls, which is what we actually want.
+:::
+
+## Conclusion
+
+Congratulations, you have completed the `Auction` contract!
+
+The [final complete code](https://github.com/sCrypt-Inc/scryptTS-examples/blob/46aaccbf5e/src/contracts/auction.ts) is as follows:
+
+```ts
+export class Auction extends SmartContract {
+    static readonly LOCKTIME_BLOCK_HEIGHT_MARKER = 500000000
+    static readonly UINT_MAX = 0xffffffffn
+
+    // The bidder's address.
+    @prop(true)
+    bidder: PubKeyHash
+
+    // The auctioneer's public key.
+    @prop()
+    readonly auctioneer: PubKey
+
+    // Deadline of the auction. Can be block height or timestamp.
+    @prop()
+    readonly auctionDeadline: bigint
+
+    constructor(auctioneer: PubKey, auctionDeadline: bigint) {
+        super(...arguments)
+        this.bidder = hash160(auctioneer)
+        this.auctioneer = auctioneer
+        this.auctionDeadline = auctionDeadline
+    }
+
+    // Call this public method to bid with a higher offer.
+    @method()
+    public bid(bidder: PubKeyHash, bid: bigint) {
+        const highestBid: bigint = this.ctx.utxo.value
+        assert(bid > highestBid, 'the auction bid is lower than the current highest bid')
+
+        // Change the address of the highest bidder.
+        const highestBidder: PubKeyHash = this.bidder
+        this.bidder = bidder
+
+        // Auction continues with a higher bidder.
+        const auctionOutput: ByteString = this.buildStateOutput(bid)
+
+        // Refund previous highest bidder.
+        const refundOutput: ByteString = Utils.buildPublicKeyHashOutput(highestBidder, highestBid)
+        let outputs: ByteString = auctionOutput + refundOutput
+
+        // Add change output.
+        if (this.changeAmount > 0) {
+            outputs += this.buildChangeOutput()
+        }
+
+        assert(hash256(outputs) == this.ctx.hashOutputs, 'hashOutputs check failed')
+    }
+
+    // Close the auction if deadline is reached.
+    @method()
+    public close(sig: Sig) {
+        // Check if using block height.
+        if (this.auctionDeadline < Auction.LOCKTIME_BLOCK_HEIGHT_MARKER) {
+            // Enforce nLocktime field to also use block height.
+            assert(this.ctx.locktime < Auction.LOCKTIME_BLOCK_HEIGHT_MARKER)
+        }
+        assert(this.ctx.sequence < Auction.UINT_MAX, 'input sequence should less than UINT_MAX')
```

Comment:
Delete here since it complicates the doc and does not change the core logic. But keep it in the final code as in the CLTV contract.